core/Tracker/Cache.php

```diff
@@ -240,14 +258,63 @@ public static function regenerateCacheWebsiteAttributes($idSites = array())
      */
     public static function deleteCacheWebsiteAttributes($idSite)
     {
-        self::getCache()->delete((int) $idSite);
+        if (self::$delegatingCacheClears) {
+            self::$delegatedClears[__FUNCTION__ . $idSite] = [__FUNCTION__, func_get_args()];
+            return;
+        }
+
+        self::getCache()->delete((int)$idSite);
     }
 
     /**
      * Deletes all Tracker cache files
      */
     public static function deleteTrackerCache()
     {
+        if (self::$delegatingCacheClears) {
+            self::$delegatedClears[__FUNCTION__] = [__FUNCTION__, []];
+            return;
+        }
+
         self::getCache()->flushAll();
     }
+
+    /**
+     * Runs `$callback` without clearing any tracker cache, just collecting which delete methods were called.
+     * After `$callback` finishes, we clear caches, but just once per type of delete/clear method collected.
+     *
+     * Use this method if your code will create many cache clears in a short amount of time (eg, if you
+     * are invalidating a lot of archives at once).
+     *
+     * @param $callback
+     */
+    public static function withDelegatedCacheClears($callback)
+    {
+        try {
+            self::$delegatingCacheClears = true;
+            self::$delegatedClears = [];
+
+            $result = $callback();
+
+            self::callAllDelegatedClears();
+
+            return $result;
+        } finally {
+            self::$delegatingCacheClears = false;
```

Comment:
would we maybe call `self::callAllDelegatedClears();` as part of finally in case there is an exception but some cache clear calls are needed?